File: 1129-longest-string-chain/1129-longest-string-chain.java
class Solution {
    public int longestStrChain(String[] words) {
        List<Set<String>> wordSets = new ArrayList<>(17);
        for(int i=0; i<17; ++i){
            wordSets.add(new HashSet<>());
        }

        for(String word : words){
            wordSets.get(word.length()).add(word);
        }

        Map<String,Integer> dp = new HashMap<>();
        int result = 1;
        for(int i=16; i>0; i--){
            if(wordSets.get(i-1).isEmpty()) continue;
            for(String word : wordSets.get(i)){
                int val = dp.getOrDefault(word,1);
                for(int j=0; j<word.length(); j++){
                    String str = word.substring(0,j) + word.substring(j+1);
                    if(wordSets.get(i-1).contains(str) && val >= dp.getOrDefault(str,1)){
                        dp.put(str,val+1);
                        result = Math.max(result, val+1);
                    }
                }
            }
        }
        return result;
    }
}

File: 1129-longest-string-chain/1129-longest-string-chain.js
/**
 * @param {string[]} words
 * @return {number}
 */
var longestStrChain = function(words) {
    let w = new Array(17).fill(0).map(()=>new Set());
    for(let i=0; i<words.length; ++i){
        w[words[i].length].add(words[i]);
    }

    let dp = new Map();
    let result = 1;
    for(let i=16; i; i--){
        if(!w[i-1].size) continue;
        for(let word of w[i]){
            let val = dp.get(word) || 1;
            for(let j=0; j<word.length; j++){
                let str = word.slice(0,j) + word.slice(j+1);
                if(w[i-1].has(str) && val >= (dp.get(str) || 1)){
                    dp.set(str,val+1);
                    result = Math.max(result, val+1);
                }
            }
        }
    }
    return result;
};